core/src/main/scala/org/http4s/blaze/pipeline/stages/SSLStage.scala

```diff
@@ -249,98 +257,50 @@ final class SSLStage(engine: SSLEngine, maxWrite: Int = 1024 * 1024)
+    if (inHandshake) {
+      handshakeQueue += DelayedWrite(data, p)
+      ()
+    } else {
+      val out = new ArrayBuffer[ByteBuffer]
+      writeLoop(data, out) match {
+        case SSLSuccess if BufferTools.checkEmpty(data) =>
+          p.completeWith(channelWrite(out)); ()
+
+        case SSLSuccess => // must have more data to write
+          channelWrite(out).onComplete {
+            case Success(_) => doWrite(data, p)
+            case Failure(t) => p.tryFailure(t); ()
+          }(serialExec)
+
+        case SSLNeedHandshake(r) =>
+          handshakeQueue += DelayedWrite(data, p)
+          val readData = takeQueuedBytes()
+          if (out.nonEmpty) { // need to write
             channelWrite(out).onComplete {
-              case Success(_) => continueWrite(data, p)
+              case Success(_) => sslHandshake(readData, r)
               case Failure(t) => p.tryFailure(t)
-            }(trampoline)
-            p.future
-
-          case SSLNeedHandshake(r) =>
-            val p = Promise[Unit]
-            handshakeQueue += DelayedWrite(data, p)
-            val readData = takeQueuedBytes()
-            if (out.nonEmpty) { // need to write
-              channelWrite(out).onComplete {
-                case Success(_) => sslHandshake(readData, r)
-                case Failure(t) => handshakeFailure(t)
-              }(trampoline)
-            } else sslHandshake(readData, r)
-
-            p.future
+            }(serialExec)
+          } else sslHandshake(readData, r)
 
-          case SSLFailure(t) => Future.failed(t)
-        }
-      }
-    }
-    logger.trace(s"${engine.##}: syncWrite completed in ${System.nanoTime - start}ns")
-    future
-  }
-
-  // Attempts to continue write requests
-  private def continueWrite(data: Array[ByteBuffer], p: Promise[Unit]): Unit = {
-    val start = System.nanoTime
-    handshakeQueue.synchronized {
-      if (inHandshake()) {
-        handshakeQueue += DelayedWrite(data, p)
-        ()
-      } else {
-        val out = new ArrayBuffer[ByteBuffer]
-        writeLoop(data, out) match {
-          case SSLSuccess if BufferTools.checkEmpty(data) =>
-            p.completeWith(channelWrite(out)); ()
-
-          case SSLSuccess => // must have more data to write
-            channelWrite(out).onComplete {
-              case Success(_) => continueWrite(data, p)
-              case Failure(t) => p.tryFailure(t); ()
-            }(trampoline)
-
-          case SSLNeedHandshake(r) =>
-            handshakeQueue += DelayedWrite(data, p)
-            val readData = takeQueuedBytes()
-            if (out.nonEmpty) { // need to write
-              channelWrite(out).onComplete {
-                case Success(_) => continueWrite(data, p)
-                case Failure(t) => p.tryFailure(t)
-              }(trampoline)
-            } else sslHandshake(readData, r)
-
-          case SSLFailure(t) => p.tryFailure(t); ()
-        }
+        case SSLFailure(t) => p.tryFailure(t); ()
```

Comment:
Why `tryFailure` here instead of `failed`?

Reply:
Old mistake. I'll change it.